pil-analyzer/src/pil_analyzer.rs

```diff
@@ -318,9 +323,48 @@ impl PILAnalyzer {
         }
     }
 
-    pub fn condense<T: FieldElement>(self) -> Analyzed<T> {
+    fn resolve_trait_impls(&mut self) -> HashMap<String, HashMap<Vec<Type>, Arc<Expression>>> {
+        let mut trait_solver = TraitsResolver::new(&self.implementations);
+
+        let mut resolve_references = |expr: &Expression| {
+            expr.all_children().for_each(|expr| {
+                if let Expression::Reference(
+                    _,
+                    Reference::Poly(
+                        reference @ PolynomialReference {
+                            type_args: Some(_), ..
+                        },
+                    ),
+                ) = expr
+                {
+                    let _ = trait_solver.resolve_trait(reference);
+                }
+            });
+        };
+
+        for (_, (_, def)) in self.definitions.iter() {
+            if let Some(FunctionValueDefinition::Expression(TypedExpression { e: expr, .. })) = def
```

Comment:
Actually for Arrays you need to do the same. Essentially the same as in infer_type, just that we don't care about the type of the expression.

Reply:
Oh and also `self.implementations` - and please add a test for that: a trait impl function that calls another trait function.